android/src/main/kotlin/com/bbflight/background_downloader/BDPlugin.kt

```diff
@@ -258,28 +261,23 @@ class BDPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
     private var pauseReceiver: NotificationRcvr? = null
     private var resumeReceiver: NotificationRcvr? = null
     private var scope: CoroutineScope? = null
-
+    private var contextID : String = ""
     /**
      * Attaches the plugin to the Flutter engine and performs initialization
      */
     override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+        //flutterPluginBinding.applicationContext
         // create channels and handler
-        backgroundChannelCounter++
-        if (backgroundChannel == null) {
-            // only set background channel once, as it has to be static field
-            // and per https://github.com/firebase/flutterfire/issues/9689 other
-            // plugins can create multiple instances of the plugin
-            backgroundChannel = MethodChannel(
-                flutterPluginBinding.binaryMessenger,
-                "com.bbflight.background_downloader.background"
-            )
-        }
+
         channel = MethodChannel(
             flutterPluginBinding.binaryMessenger, "com.bbflight.background_downloader"
         )
         channel?.setMethodCallHandler(this)
         // set context and register notification broadcast receivers
         applicationContext = flutterPluginBinding.applicationContext
+        //simple context ID generation
+        contextID = System.currentTimeMillis().toString();
```

Comment:
My main questions with the approach is here:

1. I assume that when multiple instances of your app attempt to use the plugin, that they all have a different applicationContext - can you confirm that?
2. If so, then we only need one backgroundChannel per applicationContext
3. If so, then why don't you simply store the channel under bgChannelsByContext and use the hash of the context (which I assume is also stable) as the key, so we don't have to store the context itself? And because we are told when to unregister, we could even use the applicationContext itself as the key, as we won't hold on to it after the application unregisters from the plugin
4. That way, we can look up the backgroundChannel if we have the context, and don't need to store a separate identifier (which then requires you to store that identifier with each task, which adds a second HashMap)
I'll add a few comments to illustrate what I mean, and would love your thoughts on this, as I think it simplifies matters a lot